File: src/SemanticQuarkBehaviorType.swift
public struct LinearAutoDestroy<T> {
    private var value: T?
    private var is_destroyed: Bool = false
    
    public init(val: T) {
        self.value = val
    }
    
    // Mutating garante que o estado nominal da struct mude ao ler
    public mutating func consume() throws -> T {
        if is_destroyed || value == nil {
            throw NSError(domain: "VariableAlreadyDestroyed", code: 1, userInfo: nil)
        }
        
        let extracted = value!
        self.value = nil // Zera a referência
        self.is_destroyed = true
        return extracted
    }
}

File: src/SemanticQuarkBehaviorType.scala
object LinearAutoDestroyModule:
  // Definição do Tipo Nominal Opaco
  opaque type LinearAutoDestroy[T] = Container[T]

  private class Container[T](var value: T, var isDestroyed: Boolean)

  def init[T](val: T): LinearAutoDestroy[T] = 
    Container(val, false)

  extension [T](container: LinearAutoDestroy[T])
    def consume(): T =
      if container.isDestroyed then 
        throw new IllegalStateException("VariableAlreadyDestroyed")
      
      val extracted = container.value
      container.value = null.asInstanceOf[T] // Limpa a referência do GC
      container.isDestroyed = true
      extracted
